lib/standalone.ml

```diff
@@ -0,0 +1,102 @@
+open Lwt.Infix
+
+(* Create a temporary directory. Unfortunately there is no mkdtemp(3)
+   equivalent in the standard library, so this is slightly ugly. *)
+let temp_directory =
+  let rec try_dir counter =
+    try
+      let filename = Filename.temp_file "obuilder" "" in
+      Unix.unlink filename;
+      Sys.mkdir filename 0o700;
+      filename
+    with Sys_error _ as e ->
+      if counter >= 1000 then raise e else try_dir (counter + 1)
+  in
+  try_dir 0
+
+(* Parse a manifest.json file obtained with download-frozen-image
+   (used by parse_manifest_layers and parse_manifest_env below). *)
+let parse_manifest basedir =
+  let open Yojson.Basic in
+  let open Yojson.Basic.Util in
+  let jsonfile = Filename.concat basedir "manifest.json" in
+  let json = from_file jsonfile in
+  match json |> to_list with
+  | [ json ] -> json
+  | [] ->
+    failwith "Empty manifest.json"
+  | exception Type_error (msg, _) ->
+    Fmt.failwith "Invalid format of manifest.json (%s)" msg
+  | _ ->
+    failwith "Invalid format of manifest.json"
+
+(* Parse a manifest.json file obtained with download-frozen-image, and
+   return a list of the layers of the image, in reverse order. *)
+let parse_manifest_layers basedir =
+  let open Yojson.Basic.Util in
+  let json = parse_manifest basedir in
+  let layers =
+    try
+      json |> member "Layers" |> to_list |> filter_string
+    with Type_error (msg, _) ->
+      Fmt.failwith "Invalid format of 'Layers' within manifest.json (%s)" msg
+  in
+  List.rev layers
+
+(* Parse a manifest.json file obtained with download-frozen-image, and
+   return the image environment, as a list of VAR=VALUE strings. *)
+let parse_manifest_env basedir =
+  let open Yojson.Basic in
+  let open Yojson.Basic.Util in
+  let json = parse_manifest basedir in
+  let configfile =
+    try
+      json |> member "Config" |> to_string
+    with Type_error (msg, _) ->
+      Fmt.failwith "Invalid format of 'Config' within manifest.json (%s)" msg
+  in
+  let configfile = Filename.concat basedir configfile in
+  let json = from_file configfile in
+  let env =
+    try
+      let config = json |> member "config" in
+      let env = config |> member "Env" |> to_list in
+      List.rev_map
+        (function
+         | `String s ->
+           (match Astring.String.cut ~sep:"=" s with
+            | None -> Fmt.failwith "Invalid environment stanza %S in Docker image (should be 'K=V')" s
+            | Some pair -> pair)
+         | _ -> failwith "Invalid json type in environment")
+      env
+    with Type_error (msg, _) ->
+      Fmt.failwith "Invalid format of configuration json (%s)" msg
+  in
+  env
+
+(* Extract a given layer (relative to srcdir) into destdir. *)
+let extract_layer destdir srcdir layer =
+  let srcpath = Filename.concat srcdir layer in
+  Os.sudo [ "tar" ; "-C"; destdir ; "-xpf"; srcpath ]
+
+let fetch ~log ~rootfs base =
+  try
```

Comment:
[`Lwt.catch`](https://ocsigen.org/lwt/5.5.0/api/Lwt#VALcatch) should be used instead of the `try … with` block.